### PR TITLE
refactor: Replace plain sql of getMaxTax by querybuilder and add event

### PR DESCRIPTION
**Pull request by aragon999**
### 1. Why is this change necessary?
Currently there is no way to modify the SQL which is used to retrieve the max tax. In order to extend there a replace hook is needed.

### 2. What does this change do, exactly?
Refactor the plain SQL as query builder and supply this query builder in a new `notify` event: `Shopware_Modules_Basket_GetMaxTax_QueryBuilder`.

### 3. Describe each step to reproduce the issue or behaviour.
\-

### 4. Please link to the relevant issues (if any).
\-

### 5. Which documentation changes (if any) need to be made because of this PR?
\-

### 6. Checklist

- [ ] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

### CODE REVIEW

File: engine/Shopware/Core/sBasket.php [hunk 1]
```diff
@@ -666,19 +667,41 @@ public function sInsertPremium()
     public function getMaxTax()
     {
         $sessionId = $this->session->get('sessionId');
+        if (!$sessionId) {
```
Comment (mitelg) [inline engine/Shopware/Core/sBasket.php]:
```suggestion
        if (!is_string($sessionId)) {
```

File: engine/Shopware/Core/sBasket.php [hunk 2]
```diff
@@ -666,19 +667,41 @@ public function sInsertPremium()
...
+            ))
+            ->orderBy('basket.tax_rate', 'DESC')
+            ->setMaxResults(1)
+            ->setParameter('sessionId', $sessionId)
+        ;
+
+        $this->eventManager->notify(
+            'Shopware_Modules_Basket_GetMaxTax_QueryBuilder',
+            [
+                'subject' => $this,
```
Comment (mitelg) [inline engine/Shopware/Core/sBasket.php]: please remove the reference to the class. If we want to move the method into an own service, we will get a problem, if we still need to reference the old core class :wink:

File: engine/Shopware/Core/sBasket.php [hunk 3]
```diff
@@ -666,19 +667,41 @@ public function sInsertPremium()
...
+                'sessionId' => $sessionId,
+                'queryBuilder' => $qb,
+            ]
+        );
+
+        $statement = $qb->execute();
+        \assert($statement instanceof ResultStatement);
```
Comment (mitelg) [inline engine/Shopware/Core/sBasket.php]: this should not be necessary. also asserts made some troubles in the past :grimacing:

File: engine/Shopware/Core/sBasket.php [hunk 4]
```diff
@@ -666,19 +667,41 @@ public function sInsertPremium()
...
 
+        $maxTaxId = $statement->fetchColumn();
```
Comment (mitelg) [inline engine/Shopware/Core/sBasket.php]:
```suggestion
        $maxTaxId = $statement->fetchOne();
```
`fetchColumn` is deprecated

**Review by mitelg — CHANGES_REQUESTED** [review]

File: engine/Shopware/Core/sBasket.php (hunk 3 above)
Comment (aragon999) [reply engine/Shopware/Core/sBasket.php]: I usually have a phpstan error with such things, that is why I added the assert here, since the `execute` method can return an `int`. But actually I noticed that is not very good, how I have done it, since `ResultStatement` does not has the property `fetchOne`, thus the next "issue" with this PR. Since I did not know how to replace `fetchColumn` on an object which does not has `fetchOne`.

TL;DR: I will look into it, but would be happy to get input on how to handle the ambiguous return values.

File: engine/Shopware/Core/sBasket.php [hunk 5]
```diff
@@ -666,19 +667,41 @@ public function sInsertPremium()
...
-SQL;
-        $maxTaxId = $this->connection->fetchColumn($sql, [empty($sessionId) ? session_id() : $sessionId]);
+        $qb = $this->connection->createQueryBuilder();
+        $qb
+            ->select(['product.taxID'])
+            ->from('s_order_basket', 'basket')
+            ->join('basket', 's_articles', 'product', 'product.id = basket.articleID')
+            ->where($qb->expr()->andX(
+                $qb->expr()->eq('basket.sessionID', ':sessionId'),
+                $qb->expr()->eq('basket.modus', '0')
```
Comment (mitelg) [inline engine/Shopware/Core/sBasket.php]:
```suggestion
                $qb->expr()->eq('basket.modus', \Shopware\Bundle\CartBundle\CartPositionsMode::PRODUCT)
```

File: engine/Shopware/Core/sBasket.php (hunk 3 above)
Comment (mitelg) [reply engine/Shopware/Core/sBasket.php]: this is a known issue: https://github.com/shopware/shopware/blob/5.7/.phpstan.neon#L92-L93

just omit the `$statement variable` and PHPStorm should auto-complete the right method :slightly_smiling_face:

Comment (aragon999) [reply engine/Shopware/Core/sBasket.php]: I see, I was not aware that there exists such a thing in the baseline :)

Comment (aragon999) [reply engine/Shopware/Core/sBasket.php]: Btw. I do not use PHPStorm :) but the language server works as well.

The problem was, that I forgot to run a composer install... and therefore had a too old version of `doctrine/dbal` installed. But now it should be all solved :) Thanks for the feedback.

**Review by mitelg — APPROVED** [review]

**Comment by PascalThesing** [comment]
Hey @aragon999, 

We just merge your changes.
Thank you for taking the time to create and adapt it! 

Congratulations 🎉
